spyder_kernels/customize/spydercustomize.py

```diff
@@ -862,46 +917,92 @@ def runcell(cellname, filename):
         # AttributeError --> systematically raised in Python 3
         pass
     ipython_shell = get_ipython()
+    try:
+        # Get code from spyder
+        cell_code = _frontend_request().runcell(cellname, filename)
+        if not filename:
+            filename = _frontend_request().current_filename()
+    except (CommError, TimeoutError):
+        if hasattr(ipython_shell, "cell_code"):
+            # This is the old way of transmitting informations
+            cell_code = ipython_shell.cell_code
+            del ipython_shell.cell_code
+        else:
+            _print("--Run Cell Error--\n"
+                   "Please use only with Spyder; ")
+            return
+
     namespace = _get_globals()
+    # Save previous __file__
+    namespace_has_file = '__file__' in namespace
+    if namespace_has_file:
+        namespace_filename = namespace['__file__']
     namespace['__file__'] = filename
-    try:
-        cell_code = ipython_shell.cell_code
-    except AttributeError:
-        _print("--Run Cell Error--\n"
-               "Please use only through Spyder's Editor; "
-               "shouldn't be called manually from the console")
-        return
 
     # Trigger `post_execute` to exit the additional pre-execution.
     # See Spyder PR #7310.
     ipython_shell.events.trigger('post_execute')
 
-    ipython_shell.run_cell(cell_code)
-    namespace.pop('__file__')
-    del ipython_shell.cell_code
+    try:
+        exec(compile(cell_code, filename, 'exec'), namespace)
+    except SystemExit as status:
+        # ignore exit(0)
+        if status.code:
+            ipython_shell.showtraceback(exception_only=True)
+    except BaseException:
+        # tb offset is 1 because we wrap execfile
+        ipython_shell.showtraceback(tb_offset=1)
+
+    # Restore __file__
+    if namespace_has_file:
+        namespace['__file__'] = namespace_filename
+    else:
+        namespace.pop('__file__')
 
 
 builtins.runcell = runcell
 
 
-def debugfile(filename, args=None, wdir=None, post_mortem=False):
+def cellcount(filename=None):
```

Comment:
Why is this function needed? It doesn't seem that you're using it in this PR.

Reply:
The new runcell function accept either a name or an index. I think this is better for axample in the following example:
```
print('a')
# %% Cell 1
print('b')
```
The first cell doesn't have a name, which is problematic. If I can call it by index as a fallback, I can just call `runcell(0)`. 

Now that I can call cells by index, I can write a script to run all the cells (because why not). But to do that, I need a way to know the number of cells. this is what `cellcount` does.